src/Enzo/enzo_EnzoMethodCheckGravity.cpp

```diff
@@ -0,0 +1,253 @@
+// See LICENSE_CELLO file for license and copyright information
+
+/// @file     enzo_EnzoMethodCheckGravity.cpp
+/// @author   James Bordner (jobordner@ucsd.edu)
+/// @date     2020-01-21
+/// @brief    Implementation of the EnzoMethodCheckGravity method
+
+#include "enzo.hpp"
+  
+//----------------------------------------------------------------------
+
+EnzoMethodCheckGravity::EnzoMethodCheckGravity ( std::string particle_type ) throw() 
+  : Method (),
+    particle_type_(particle_type)
+{
+  cello::simulation()->new_refresh_set_name(ir_post_,name());
+
+  Refresh * refresh = cello::refresh(ir_post_);
+  refresh->add_all_particles();
+}
+
+//----------------------------------------------------------------------
+
+void EnzoMethodCheckGravity::compute ( Block * block) throw()
+{
+  if (block->is_leaf()) {
+
+    Particle particle (block->data()->particle());
+    Field    field    (block->data()->field());
+
+    // initialize CheckGravity particle type and position attributes
+
+    const int it = particle.type_index(particle_type_);
+
+    const int ia_x = particle.attribute_index(it,"x");
+    const int ia_y = particle.attribute_index(it,"y");
+    const int ia_z = particle.attribute_index(it,"z");
+
+    const int dp =  particle.stride(it,ia_x);
+
+    const int rank = cello::rank();
+
+    // get velocity field arrays
+
+    union { float * vxa4; double * vxa8; };
+    union { float * vya4; double * vya8; };
+    union { float * vza4; double * vza8; };
+  
+    // NOTE: union so v?a4 also initialized
+    
+    vxa8 = (rank >= 1) ? (double *) field.values("velocity_x") : NULL;
+    vya8 = (rank >= 2) ? (double *) field.values("velocity_y") : NULL;
+    vza8 = (rank >= 3) ? (double *) field.values("velocity_z") : NULL;
+
+    int mx,my,mz;
+    field.dimensions(0,&mx,&my,&mz);
+    int nx,ny,nz;
+    field.size(&nx,&ny,&nz);
+    int gx,gy,gz;
+    field.ghost_depth(0,&gx,&gy,&gz);
+
+    double xm,ym,zm;
+    double xp,yp,zp;
+    block->lower(&xm,&ym,&zm);
+    block->upper(&xp,&yp,&zp);
+
+    const double hx = (xp-xm)/nx;
+    const double hy = (yp-ym)/ny;
+    const double hz = (zp-zm)/nz;
+
+    double dt = block -> dt();
+
+    // Get velocity precision: ASSUMES precision same for all axes and is single or double
+    
+    const bool is_single = (field.precision (field.field_id("velocity_x")) == precision_single);
```

Comment:
What happens if compiled with quadruple prec? Should there be a check here (or is there already somewhere else I missed) to make sure the code fails with an informative error message if quadruple is used?